'Source directory does not exist' building Linux with CMake 3.10

使用 flutter run -d linux 出现了下面的问题

Launching lib/main.dart on Linux in debug mode...
CMake Error: The source directory "/home/mi/hello/build/linux/debug" does not exist.
Specify --help for usage, or press the help button on the CMake GUI.
Exception: Unable to generate build files

过程：
1, mkdir hello && cd hello
2, flutter create . && flutter pub get 
3, flutter  run -d linux 

flutter version:
[✓] Flutter (Channel master, 1.19.0-2.0.pre.83, on Linux, locale zh_CN.UTF-8)
    • Flutter version 1.19.0-2.0.pre.83 at /home/mi/flutter
    • Framework revision 8abf0a6d8c (5 小时前), 2020-05-16 15:07:34 -0700
    • Engine revision aaf9e79f1d
    • Dart version 2.9.0 (build 2.9.0-9.0.dev 2676764792)
    • Pub download mirror https://pub.flutter-io.cn
    • Flutter download mirror https://storage.flutter-io.cn

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at /home/mi/Android/Sdk
    • Platform android-29, build-tools 29.0.3
    • Java binary at: /home/mi/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_212-release-1586-b4-5784211)
    • All Android licenses accepted.

[✓] Linux toolchain - develop for Linux desktop
    • clang version 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
    • cmake version 3.10.2
    • ninja version 1.8.2

[✓] Android Studio (version 3.6)
    • Android Studio at /home/mi/android-studio
    • Flutter plugin version 45.1.1
    • Dart plugin version 192.7761
    • Java version OpenJDK Runtime Environment (build
      1.8.0_212-release-1586-b4-5784211)

[✓] VS Code (version 1.45.0)
    • VS Code at /usr/share/code
    • Flutter extension version 3.10.1

[✓] Proxy Configuration
    • HTTP_PROXY is set
    • NO_PROXY is localhost,127.0.0.0/8,::1,127.0.0.1
    • NO_PROXY contains 127.0.0.1
    • NO_PROXY contains localhost

[✓] Connected device (1 available)
    • Linux • Linux • linux-x64 • Linux

• No issues found!

--- a/packages/flutter_tools/lib/src/linux/build_linux.dart
+++ b/packages/flutter_tools/lib/src/linux/build_linux.dart
@@ -81,20 +81,20 @@ Future<void> buildLinux(
 Future<void> _runCmake(String buildModeName, Directory sourceDir, Directory buildDir) async {
   final Stopwatch sw = Stopwatch()..start();
 
+  await buildDir.create(recursive: true);
+
   final String buildFlag = toTitleCase(buildModeName);
   int result;
   try {
     result = await processUtils.stream(
       <String>[
         'cmake',
-        '-S',
-        sourceDir.path,
-        '-B',
-        buildDir.path,
         '-G',
         'Ninja',
         '-DCMAKE_BUILD_TYPE=$buildFlag',
+        sourceDir.path,
       ],
+      workingDirectory: buildDir.path,
       environment: <String, String>{
         'CC': 'clang',
         'CXX': 'clang++'

--- a/packages/flutter_tools/test/commands.shard/hermetic/build_linux_test.dart
+++ b/packages/flutter_tools/test/commands.shard/hermetic/build_linux_test.dart
@@ -82,28 +82,38 @@ void main() {
     projectTemplateVersionFile.writeAsStringSync(templateVersion.toString());
   }
 
-  // Returns the command list matching the build_linux call to cmake.
-  List<String> cmakeCommand(String buildMode) {
-    return <String>[
-      'cmake',
-      '-S',
-      '/linux',
-      '-B',
-      'build/linux/$buildMode',
-      '-G',
-      'Ninja',
-      '-DCMAKE_BUILD_TYPE=${toTitleCase(buildMode)}',
-    ];
+  // Returns the command matching the build_linux call to cmake.
+  FakeCommand cmakeCommand(String buildMode, {void Function() onRun}) {
+    return FakeCommand(
+      command: <String>[
+        'cmake',
+        '-G',
+        'Ninja',
+        '-DCMAKE_BUILD_TYPE=${toTitleCase(buildMode)}',
+        '/linux',
+      ],
+      workingDirectory: 'build/linux/$buildMode',
+      onRun: onRun,
+    );
   }
 
-  // Returns the command list matching the build_linux call to ninja.
-  List<String> ninjaCommand(String buildMode) {
-    return <String>[
-      'ninja',
-      '-C',
-      'build/linux/$buildMode',
-      'install',
-    ];
+  // Returns the command matching the build_linux call to ninja.
+  FakeCommand ninjaCommand(String buildMode, {
+    Map<String, String> environment,
+    void Function() onRun,
+    String stdout = '',
+  }) {
+    return FakeCommand(
+      command: <String>[
+        'ninja',
+        '-C',
+        'build/linux/$buildMode',
+        'install',
+      ],
+      environment: environment,
+      onRun: onRun,
+      stdout: stdout,
+    );
   }
 
   testUsingContext('Linux build fails when there is no linux project', () async {
@@ -165,8 +175,8 @@ void main() {
   testUsingContext('Linux build invokes CMake and ninja, and writes temporary files', () async {
     final BuildCommand command = BuildCommand();
     processManager = FakeProcessManager.list(<FakeCommand>[
-      FakeCommand(command: cmakeCommand('release')),
-      FakeCommand(command: ninjaCommand('release')),
+      cmakeCommand('release'),
+      ninjaCommand('release'),
     ]);
 
     setUpMockProjectFilesForBuild();
@@ -186,7 +196,7 @@ void main() {
     final BuildCommand command = BuildCommand();
     setUpMockProjectFilesForBuild();
     processManager = FakeProcessManager.list(<FakeCommand>[
-      FakeCommand(command: cmakeCommand('release'), onRun: () {
+      cmakeCommand('release', onRun: () {
         throw ArgumentError();
       }),
     ]);
@@ -205,8 +215,8 @@ void main() {
     final BuildCommand command = BuildCommand();
     setUpMockProjectFilesForBuild();
     processManager = FakeProcessManager.list(<FakeCommand>[
-      FakeCommand(command: cmakeCommand('release')),
-      FakeCommand(command: ninjaCommand('release'), onRun: () {
+      cmakeCommand('release'),
+      ninjaCommand('release', onRun: () {
         throw ArgumentError();
       }),
     ]);
@@ -225,10 +235,10 @@ void main() {
     final BuildCommand command = BuildCommand();
     setUpMockProjectFilesForBuild();
     processManager = FakeProcessManager.list(<FakeCommand>[
-      FakeCommand(command: cmakeCommand('debug')),
-      FakeCommand(
-        command: ninjaCommand('debug'),
-        stdout: 'STDOUT STUFF',),
+      cmakeCommand('debug'),
+      ninjaCommand('debug',
+        stdout: 'STDOUT STUFF',
+      ),
     ]);
 
     await createTestCommandRunner(command).run(
@@ -247,9 +257,8 @@ void main() {
     final BuildCommand command = BuildCommand();
     setUpMockProjectFilesForBuild();
     processManager = FakeProcessManager.list(<FakeCommand>[
-      FakeCommand(command: cmakeCommand('debug')),
-      FakeCommand(
-        command: ninjaCommand('debug'),
+      cmakeCommand('debug'),
+      ninjaCommand('debug',
         environment: const <String, String>{
           'VERBOSE_SCRIPT_LOGGING': 'true'
         },
@@ -273,8 +282,8 @@ void main() {
     final BuildCommand command = BuildCommand();
     setUpMockProjectFilesForBuild();
     processManager = FakeProcessManager.list(<FakeCommand>[
-      FakeCommand(command: cmakeCommand('debug')),
-      FakeCommand(command: ninjaCommand('debug')),
+      cmakeCommand('debug'),
+      ninjaCommand('debug'),
     ]);
 
 
@@ -292,8 +301,8 @@ void main() {
     final BuildCommand command = BuildCommand();
     setUpMockProjectFilesForBuild();
     processManager = FakeProcessManager.list(<FakeCommand>[
-      FakeCommand(command: cmakeCommand('profile')),
-      FakeCommand(command: ninjaCommand('profile')),
+      cmakeCommand('profile'),
+      ninjaCommand('profile'),
     ]);
 
     await createTestCommandRunner(command).run(
@@ -310,8 +319,8 @@ void main() {
     final BuildCommand command = BuildCommand();
     setUpMockProjectFilesForBuild();
     processManager = FakeProcessManager.list(<FakeCommand>[
-      FakeCommand(command: cmakeCommand('release')),
-      FakeCommand(command: ninjaCommand('release')),
+      cmakeCommand('release'),
+      ninjaCommand('release'),
     ]);
     fileSystem.file('lib/other.dart')
       .createSync(recursive: true);
@@ -396,8 +405,8 @@ set(BINARY_NAME "fizz_bar")
     final BuildCommand command = BuildCommand();
     setUpMockProjectFilesForBuild();
     processManager = FakeProcessManager.list(<FakeCommand>[
-      FakeCommand(command: cmakeCommand('release')),
-      FakeCommand(command: ninjaCommand('release')),
+      cmakeCommand('release'),
+      ninjaCommand('release'),
     ]);
 
     await createTestCommandRunner(command).run(

